decidim-core/lib/decidim/exporters/pdf.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+require "wicked_pdf"
+
+module Decidim
+  module Exporters
+    # Exports a PDF using the provided hash, given a collection and a
+    # Serializer. This is an abstract class that should be inherited
+    # to create PDF exporters, with each PDF exporter class setting
+    # the desired template, layout and orientation.
+    #
+    class PDF < Exporter
+      # Public: Exports a PDF version of the collection by rendering
+      # the template into html and then converting it to PDF.
+      #
+      # Returns an ExportData instance.
+      def export
+        html = controller.render_to_string(
+          template: template,
+          layout: layout,
+          locals: { collection: participants }
+        )
+
+        document = WickedPdf.new.pdf_from_string(html, orientation: orientation)
```

Comment:
Generating PDF documents during request time causes memory and response time issues.
The document should be generated in a background job and delivered to the user via email or other means please.

Reply:
PDFs are generated as background jobs except for single responses, which shouldn't need much time to be generated. It is kind of uncomfortable for the user to have to check their email just to see one response.

If you still consider individual responses exported to PDF should be processed in background, we could set a config variable to enable or disable individual response exports.

Reply:
Hi (github doesn't allow me to mention you),
PDF generation is causing problems in some deployments. Specially when there are specific stylings, like this one. So we prefer not to have any config option, but to always generate resource consuming tasks in the background.

If you don't like the email, we can accept to open a new window with a message and a spinner that polls the server every 15secs until the file is generated, and then return it to the user.

Reply:
Hi @tramuntanal! I'll check my config to see if there's something about mentions that I have disabled.

Regarding the PDF generation, I've just commited changes so individual answers are also exported in background. I wanted to mention that one test is failing and I couldn't get it to work, I think it's related to using `redirect_back` in the controller.

Could please help me get the test to pass if you have any idea about why it's failing? Thanks!

```
  1) Decidim::Surveys::Admin::SurveysController export_response redirects to show template with a flash notice message
     Failure/Error: get :export_response, params: params
     
     ActionController::UrlGenerationError:
       No route matches {}, missing required keys: [:component_id, :initiative_slug]
     # /home/veri/Platoniq/survey-answers-decidim/decidim-forms/app/helpers/decidim/forms/admin/concerns/has_questionnaire_answers_url_helper.rb:30:in `questionnaire_participant_answers_url'
     # /home/veri/Platoniq/survey-answers-decidim/decidim-forms/app/controllers/decidim/forms/admin/concerns/has_questionnaire_answers.rb:60:in `export_response'
     # /home/veri/Platoniq/survey-answers-decidim/decidim-core/app/controllers/concerns/decidim/use_organization_time_zone.rb:21:in `use_organization_time_zone'
     # /home/veri/Platoniq/survey-answers-decidim/decidim-core/app/controllers/concerns/decidim/locale_switcher.rb:24:in `switch_locale'
     # ./spec/controllers/decidim/surveys/admin/surveys_controller_spec.rb:74:in `block (3 levels) in <module:Admin>'
```

Reply:
The problem here is that the function `questionnaire_participant_answers_url` doesn't find the proper route for the component, probably a initiative has mounted a survey too and tries to use it.
One solution would be to override that function in the `surveys_controller.rb` by sending the specific mounted route engine:

```ruby
        def questionnaire_for
          survey
        end

        # specify where to redirect after exporting an answer
        def questionnaire_participant_answers_url(session_token)
          Decidim::EngineRouter.admin_proxy(survey.component).show_survey_path(session_token: session_token)
        end
```

Reply:
Thank you @microstudi, I'll make the change.